scripts/fireupgrade.sh

```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
+
+# This script should only handle upgrade, nothing else
+
+if [[ -e "/home/pi/.firewalla/config/.no_auto_upgrade" ]]; then
+  exit 0
+fi
+
+cd /home/pi/firewalla
+cd .git
+sudo chown -R pi *
+cd ..
+branch=$(git rev-parse --abbrev-ref HEAD)
+
+GITHUB_STATUS_API=https://status.github.com/api.json
+
+for i in `seq 1 5`; do
+    HTTP_STATUS_CODE=`curl -s -o /dev/null -w "%{http_code}" $GITHUB_STATUS_API`
+    if [[ $HTTP_STATUS_CODE == "200" ]]; then
+      break
+    fi
+    sleep 1
+done
+
+
+# continue to try upgrade even github api is not successfully.
+# very likely to fail
+
+echo "upgrade on branch $branch"
+
+(sudo -u pi git fetch origin $branch && sudo -u pi git reset --hard FETCH_HEAD) || exit 1
```

Comment:
use "git reset --hard FETCH_HEAD" instead of "git reset --hard origin/$branch"

FETCH_HEAD is more accurate than origin/$branch. I had several times that origin/$branch is not latest as FETCH_HEAD (not 100% reproducible), which might be an issue why user's pi didn't get upgraded successfully.